Add base32check article (#1)
espadrine/espadrine.github.com: blog/src/a-base32-checksum.md

# A base32 checksum

_(Cue: in the voice of a badly-acted TV commercial.)_

> Do you always mistakenly wire money to or from the wrong account?
> I used to do that all day, everyday!
>
> But then, I found base32check, and my life turned upside-down.
> base32check is there for me on my daily purchases, watching out for my

> > erroneous transcriptions online and offline! And now, for some reason,

Reviewer: What's a transcription? Maybe that's a bit too jargon-y

Author: Agreed! Especially going in, when you don't know what the subject is. I think `typing errors` is better than `erroneous transcriptions`.

Reviewer: Yes!

> I smell of roses and there are sparkling AfterEffects™ all around.
>
> Download the base32check whitepaper. Symptoms may include an accelerated heart
> rate, a shortness of breath, and acute depression. Please consult your
> healthcare professional in case of severe brain injury.

> ## That was weird, let's be boring instead

Reviewer: Loved the intro and how that line breaks it and moves on 💯

    luhnValidity.textContent = luhnValid(value) ?  'Valid' : 'Invalid';
    const digit = luhnGen(value);
    luhnDigit.textContent = (digit !== digit) ? 'removal of non-digits' : digit;
  };
  luhnInput.addEventListener('input', updateWidget);
  updateWidget();
}
addEventListener('DOMContentLoaded', initLuhnWidget);
</script>

Whether a single check digit could also detect all adjacent transpositions
was an open question, until 1969, when **[Verhoeff][]** designed another
algorithm that did just that.
Furthermore, he gathered the results of a study into a thorough analysis
of the causes of human errors.

Meanwhile, the International Standards Organization published
[ISO/IEC 7064][] in 1983. It includes algorithms for single- and
double-digit checksums of numerical and alphanumerical inputs, with the goal
to detect all single substitutions and most adjacent transpositions.

They intended it for use in future standards.
For instance, **[IBAN][ISO 13616]** (the International Bank Account Number)
needed a checksum to lower the risk of accidentally transfering funds to the
wrong account number.

Surprisingly, the IBAN system, which supports alphanumeric characters,

> relied on MOD 97-10, which was designed only for numeric inputs.

Reviewer: Is IBAN's ISO 1316 using MOD 97-10? I'm a bit confused when I read this part, as to whether the IBAN standard did or did not follow ISO/IEC 7064, and whether ISO/IEC 7064 did introduce MOD 97-10 checks

Author: Good point! This is likely clearer:

```markdown
They intended it for use in future standards.
For instance, **[IBAN][ISO 13616]** (the International Bank Account Number),
published in 1997, needed a checksum to lower the risk of accidentally
transfering funds to the wrong account number.

The designers created IBAN as a sequence of alphanumeric characters.
Surprisingly, for the check digits, they relied on [ISO/IEC 7064][]’s MOD 97-10,
which was designed only for numeric inputs.
```

Reviewer: Much clearer to me :)

Our analysis indicates that this was a BAD MOVE™.

> A root promise of MOD 97-10 as specified in [ISO/IEC 7064][] is to detect

Reviewer: Oh ok, it gets clearer here but maybe could have been a bit more obvious higher up 👍

3. **Contain enough space to scale globally**. We don't want to run out of
   account numbers as the population grows. This, along with the previous
   constraint, means that we need to encode as many bits of information as
   we can within each character. base64 maybe? That'd be 6 bits per char.
4. **Minimize human transcription errors**. base64 is notably poor in this
   respect, because of the 0/O and I/l/1/i visual confusion. On the other
   hand, base32 has more information density than just digits, but none of
   the ambiguity. Finally, while base64 contains a dash or a dot that can
   be cut off by text processors, base32 will be selected as a single word
   even when double-clicking.
5. **Allow early detection of errors**. We don't want to inform the user
   of their mistake asynchronously. So the algorithm should be easy to
   implement on the user-interface.

Obviously, I am not the first to look into this. Notably, [Bitcoin][base58]
created a solution for this. How does it work, and how does it fare?

In Bitcoin, the account number is roughly a (cryptographic) hash of the user's
public key, cut and concatenated with a checksum that also uses a cryptographic
hash.

> The account number is usually represented in text as [base58][]. It is very

Reviewer: Here I lost context (the fact that you were talking about Bitcoin account numbers). Might be worth making it more obvious that you are talking about bitcoin in this paragraph and the one below about its Checksum. I see two ways of doing this, either intro'ing each paragraph in a more obvious way ("Bitcoin's account numbers" / "Bitcoin's checksum"), or heading all of those paragraph with a title (e.g: "Bictoin's solution... and its shortcomings")

Author: I like re-intro'ing each paragraph :)

similar to [base64][], except that it removes the 0/O and l/I ambiguity, but
still has the i/1, Z/7, B/8 and S/5 confusion. Besides, transcribing or
communicating mixed-case letters is tedious. Each letter needs mentioning
whether it is uppercase or lowercase. That is why [fourwordsalluppercase][] is a
great WiFi password.

The checksum uses roughly 5 characters (four bytes of [base58][] is 32 ÷
log2(58)). It is fairly big, which at least ensures that there won't be
mistakes… unless the account number goes unchecked. If that is the case, neither
the sender nor the recipient may use the funds transferred.

Sadly, the necessity of a bignum library for base58 decoding, and SHA256 as the
checksum algorithm, makes verifying the checksum laborious to implement.
Few user interfaces embed a validator as a result.

Finally, even if we wanted to replace the checksum with an easier-to-use check
digit, it is not immediately easy to design a base58 check character.

[base32][] avoids many of these inconveniences, especially in lowercase, where

> the varied heights are easy to distinguish and where there is no S/5 and Z/7

Reviewer: ```suggestion
the varied heights are easy to distinguish and where there are no S/5 and Z/7
```

> ambiguity.

Reviewer: ```suggestion
ambiguities.
```

However, there isn't a particularly good, dedicated, check character algorithm
for it — yet.

Let's create one.

## base32check1

Luhn, Verhoeff, Damm — some pretty nifty designs are already on the shelves.
Sadly, they mostly focus on digits, which have poor information capacity (3.3
bits per char).

> While they can be extended to larger alphabets, that is not trivial nor

Reviewer: ```suggestion
While they can be extended to larger alphabets, it is not trivial nor
```

standard. (Ah, the joy of tweaking the Verhoeff parameters.) Worse, they are no
longer the coolest kids on the block.

I found a 2016 [paper][Chen 2016] referencing a 2014 one (which I couldn't get
my hands on. Curse the IEEE overseers and their labyrinthic subscription

> models! _Please pay $84.50 membership and the paper is $14.95, oh and would you

Reviewer: ```suggestion
models! _Please pay a $84.50 membership fee, and the paper is $14.95. Oh, and would you
```

> mind donating because IEEE helps improve the human condition? Also good news, we

Reviewer: :trollface:

There are two possible cardinals: *prime numbers* and *prime powers*. Prime
cardinals are neat, because operations are scalar. On the other hand, additions
and multiplications on prime power cardinal fields are matricial.

I first wanted a single-character checksum. In order to use all the bits that
this checksum had at its disposal, I needed it to support the full base32
alphabet. That meant the cardinal needed to be 32. Sadly, 32 is not a prime; so
it needed to be a prime power. If 2 is the prime, 2⁵ is the cardinal.

Since the cardinal is not a prime, P is a matrix. Here, its size is 5×5.
To build it, we can pick any order-5 [primitive polynomial][]; I chose 1+x²+x⁵.
To get the companion matrix, set a diagonal of 1s, and set the rightmost column
with the polynomial coefficients, starting with x⁰:

    ⎛ 0 0 0 0 1 ⎞
    ⎜ 1 0 0 0 0 ⎟
    ⎜ 0 1 0 0 1 ⎟
    ⎜ 0 0 1 0 0 ⎟
    ⎝ 0 0 0 1 0 ⎠

Each aᵢ term is a representation of a base32 character as a GF(2⁵) (finite field
of order 2⁵) polynome: a is 0, which we represent as (0 0 0 0 0); b is 1 or (0 0
0 0 1); f is 1+x² or (0 0 1 0 1).

Validating a check character is straightforward: compute Σ aᵢ·Pⁱ; verify that

> the result is a zero vector. It does involve writing custom matrix

Reviewer: ```suggestion
the result is a zero vector. It does involve writing a custom matrix
```

Should be either `a custom matrix multiplication` or `custom matrix multiplications`, I think

> multiplication and addition primitives to ensure that they are all computed

Reviewer: something sounds weird to me here, looks like `addition` is used as a verb - did you mean `writing custom matrices, multiplying and summing up primitives to ensure...`? 
(I'm totally out of depth technically for this section of the article to make sense to me so please be kind if my comment appears overly naive or off-topic 💃 )

Author: The way it is meant to be parsed is `writing (custom matrix (multiplication and addition)) primitives`. It uses the programmatic meaning of "primitive" (a simple operation relied upon to compute larger algorithms), although it is confusing here, since we use the mathematical meaning (which is very different) in a previous paragraph.

`writing custom matrix operations` is simpler.

Reviewer: Got it, makes sense now that you lay it out for me - maybe readers who are more experienced in these types of technical posts will have no problem parsing this (but the simpler version works better for me) 👍

modulo the cardinal.

Unsurprisingly, the performance is not all that great:

    $ node test/perf
    10000 runs.
    base32check1: 243.500ms

There is a [neat trick][ffcomp] to speed up computation on a finite field where
the cardinal is a power of 2: instead of implementing vectors as arrays, use
integers whose binary sequence is the list of polynome coefficients. For
instance, 1+x² (which represents f) is `0b00101`, or 5.

Then **addition is XOR**, and matrix multiplication is a conditional XOR on the
successive bits of the current row. With aᵢ and bᵢ vectors of size 5:

    ⎛ a₀ ⎞   ⎛ b₀ ⎞   ⎛ Σ a₀[i]×bᵢ ⎞
    ⎜ a₁ ⎟   ⎜ b₁ ⎟   ⎜ Σ a₁[i]×bᵢ ⎟
    ⎜ a₂ ⎟ × ⎜ b₂ ⎟ = ⎜ Σ a₂[i]×bᵢ ⎟
    ⎜ a₃ ⎟   ⎜ b₃ ⎟   ⎜ Σ a₃[i]×bᵢ ⎟
    ⎝ a₄ ⎠   ⎝ b₄ ⎠   ⎝ Σ a₄[i]×bᵢ ⎠

It is 7× faster.

    $ node test/perf
    10000 runs.
    base32check1: 34.120ms

Now, how do you compute the check character?

First, compute S = Σ aᵢ·Pⁱ for the data you have.
Then, we need to solve S + c·P<sup>n+1</sup> = 0 for c.

Let's compute c·P<sup>n+1</sup> = -S.
Opposites are equal in GF(2ᵏ), so -S = S.

> The second insight is that, since a primitive element is a generator of

Reviewer: Are you talking about the second insight of the 2016 paper?

Author: Ah, no. You need two insights to solve the equation. The first trick is that computing opposites is trivial because it is an identity operation; the second is this one.

I will switch to this formulation:

```markdown
Now, we need to inverse P<sup>n+1</sup>.
To do that efficiently, we need the following insight: since a primitive element […]
```

  // Therefore:      a * a^(cardinal-2) = 1
  // Here we have:   a = primitive^(n+1)
  // Hence:          code = opposite * primitive^((cardinal-2)*(n+1))
  let exp = (cardinal-n-2) % (cardinal - 1);
  exp = (exp < 0)? exp + cardinal: exp;
  const inverse = primitivePowers[exp];
  const code = matMul([opposite], inverse)[0];
  //console.log(`opposite ${opposite}\texp ${exp}\tinverse ${JSON.stringify(inverse)}\tcode ${code}`);
  return toBase32Char(code);
}

function validate(payload) {
  return compute(payload) === 'a';
}

function init() {
  const widget = document.getElementById('base32check1Widget');
  const input = widget.getElementsByClassName('input')[0];
  const validity = widget.getElementsByClassName('validity')[0];
  const checksum = widget.getElementsByClassName('checksum')[0];
  const updateWidget = () => {
    const value = input.value;
    validity.textContent = validate(value) ?  'Valid' : 'Invalid';
    const check = compute(value);
    checksum.textContent = (check !== check) ? 'removal of non-base32' : check;
  };
  input.addEventListener('input', updateWidget);
  updateWidget();
}
addEventListener('DOMContentLoaded', init);
})();
</script>

## base32check2

> I chose to also design a 2-character checksum alongside, on the off-chance that

Reviewer: Alongside what? 
```suggestion
I chose to also design a 2-character checksum alongside the first, on the off-chance that
```

> the first was not strong enough. There are 2<sup>2×5</sup> = 1024 possible

Reviewer: What makes you doubt that the first isn't strong enough?

Author: I am making a single (5-bit) character check compete with a double character check (6.6 bits total). I needed to give better results with less space. Even though it ended up working out, it was a bit ambitious.

I will clarify this initial estimate.

combinations of two characters of base32. Using the same design as above, it
required working on 10×10 matrices. I chose not to do that for the following
reasons:

> - The initialization matrix starts to look scary like the Verhoeff one!

Reviewer: ```suggestion
- The initialization matrix starts to look as scary as Verhoeff's!
```

    return this.expected;
  }

  validate() {
    const errors = [];
    if (typeof this.alphabet !== 'string') {
      errors.push({code: 'invalid_type', message: 'test alphabet is not a string'});
    } else if (!isValidAlphabet(this.alphabet)) {
      errors.push({code: 'invalid_type', message: 'test alphabet is unknown'});
    }
    if (typeof this.bits !== 'number') {
      errors.push({code: 'invalid_type', message: 'test bits are not a number'});
    }
    if (typeof this.expected !== 'string') {
      errors.push({code: 'invalid_type', message: 'expected test input is not a string'});
    } else if (this.expected.length > 200) {
      errors.push({code: 'invalid_type', message: 'expected test input is too long'});
    }
    if (errors.length > 0) { return errors; }
    const expectedSize = stringSizeToEncode(this.bits, this.alphabet);
    if (expectedSize !== this.expected.length) {
      errors.push({code: 'invalid_type', message: 'expected test input does not match alphabet and bits'});
    }
    return errors;
  }
}

  const submission = new Submission({
    user: '',
    total: Object.fromEntries(alphabetNames.map(n => [n, 0])),
    errors: [],
  });

>   const submissionBatchSize = 50;

Reviewer: Why 50? (Just curious, maybe more people would go through the experiment and reach its conclusion if the sample size took them less time to complete)

Also, once you do reach 50, you still have the text <input> visible, while anything you enters in it doesn't get submitted, so it's a bit confusing at that point. It'd be great to remove the input when the experiment is over, and maybe even replace it with a conclusion as to which algorithm performed better on my sample vs. the whole user entries sample?

Also, maybe the test would be more realistic if the information to copy were not immediately above the input, but to the left of it. Having it right above the input makes it too easy to visually compare the two, which won't be the case in realistic IBAN entry use-cases. What do you think?

https://www.loom.com/share/1052820091f14fe09ea8511dee1d308f

Author: Good points!

The batch limit unlocks the overall study results as incentive to hit that mark. Submissions are persisted on the database server on every entry. Even after 50, the new entries are submitted, and the table updated. This can be made clearer; I will add an explanation when we hit 50.
